### DISCUSSION TITLE
Newbie: Including 3rd party modules for APK

### DISCUSSION THREAD

**Discussion by bassemb** [beeware/toga#1312] [Q&A]
Hi, sort of a noob here. Trying to learn app developing, so I figured I'd give it a go by trying to make a simple app that downloads Fb videos. I've got all the logic and the GUI running properly on Windows, but after creating, building and running through Briefcase, the app kept crashing on startup on my Android phone. With some narrowing down, I found that importing other modules caused the crash every time. How do I go about fixing this issue? I'm importing the modules datetime, os, requests and re. Again, I've been coding for 2 months now so it's all very hazy to me.

**Comment by freakboy3742** [comment]
Welcome!

I'm guessing that you're missing a step: adding the third party package in your `pyproject.toml`. Briefcase builds a complete standalone environment for your app, and part of that requires that you declare all the libraries that are part of your app. It's not enough to just `pip install` the library - you have to add the package name (and version, if required) to the `requires` list.

Once you've done that, you can either re-create the app, or run `briefcase update -d` (to update dependencies), then rebuild and re-run.

**Reply by bassemb** [reply]
Thanks for the super quick reply. I'll give that a go tomorrow when I get back on it! :)

**Reply by bassemb** [reply]
> Welcome!
> 
> I'm guessing that you're missing a step: adding the third party package in your `pyproject.toml`. Briefcase builds a complete standalone environment for your app, and part of that requires that you declare all the libraries that are part of your app. It's not enough to just `pip install` the library - you have to add the package name (and version, if required) to the `requires` list.
> 
> Once you've done that, you can either re-create the app, or run `briefcase update -d` (to update dependencies), then rebuild
 and re-run.

Hey, I managed to get the app running, but I'm facing two other issues:
1- I have a button that adds other buttons working properly on Windows, but not on my Android. The child buttons just don't show up.
2- For some reason briefcase now isn't updating the apps anymore. The processes are completing but the app itself isn't changing from its last state when installed again on the phone. 

Sorry for all the questions, all of this is so interesting and fun yet so incomprehensible for someone of my level. What's the quickest way to debug issues like that? Thanks

Here's the console output for create and build

[build.txt](https://github.com/beeware/toga/files/6901352/build.txt)
[create.txt](https://github.com/beeware/toga/files/6901354/create.txt)

**Reply by freakboy3742** [reply]
Regarding the buttons not showing up on Android - that's likely a bug. The Android backend is one of the newest additions to Toga, and getting layouts to draw and redraw consistently is one of the hardest pieces to get right on any platform. If you're able to narrow your problem down to a simple example that reproduces the behavior, and log that as a ticket on the Toga repository, that would be very helpful.

Having updates not apply is a different problem, and definitely a concerning one. I've heard similar reports from others; but I've never been able to reproduce it without deliberately missing out an `update` or `build` step. Again, if you're able to provide a consistent set of instructions to reproduce the problem, and log those as a ticket on the Briefcase repository, it would be very helpful.

As for the general question about debugging - I'm not sure what to suggest. The best debugging tools I know of are copious print statements, and changing the background color of widgets to check the bounds of layout geometries.
